File: features/pdf_scanning/service.py
"""
Services for PDF Scanning Feature
"""

import os
import logging
import uuid
from datetime import datetime
import json
from pathlib import Path
from pdf_processor import DocumentProcessor # Import the new processor
from config.configuration import document_processor_config # Import the config for the processor

logger = logging.getLogger(__name__)

# In-memory document storage (in a real app, this would be in a database)
documents = {}

# Instantiate the DocumentProcessor (consider making this a singleton or managed instance later)
# Pass the specific configuration section for the processor
document_processor = DocumentProcessor(config=document_processor_config)

def process_pdf_document(file_path):
    """
    Process a PDF document
    
    Args:
        file_path: Path to the PDF file
        
    Returns:
        dict: Processing result
    """
    logger.info(f"Processing PDF document: {file_path}")
    
    # Generate document ID from filename
    file_name = os.path.basename(file_path)
    document_id = f"{uuid.uuid4()}" # Generate a unique ID for this processing instance

    try:
        # Use the unified DocumentProcessor
        result = document_processor.process_document(file_path)

        if 'error' in result:
            # Processing failed
            logger.error(f"DocumentProcessor failed for {file_path}: {result['error']}")
            # Store minimal failure info
            documents[document_id] = {
                'id': document_id,
                'file_path': file_path,
                'file_name': file_name,
                'uploaded_at': datetime.now().isoformat(),
                'processing_result': {'error': result['error']},
                'metadata': result.get('metadata', {}) # Include metadata if available
            }
            return {
                'document_id': document_id,
                'status': 'error',
                'message': f"Error processing document: {result['error']}",
                'metadata': result.get('metadata', {})
            }
        else:
            # Processing succeeded
            logger.info(f"DocumentProcessor succeeded for {file_path}")
            # Store the full result
            documents[document_id] = {
                'id': document_id,
                'file_path': file_path,
                'file_name': file_name,
                'uploaded_at': datetime.now().isoformat(),
                'processing_result': result, # Store the detailed result
                'metadata': result.get('metadata', {}) # Extract top-level metadata
            }
            return {
                'document_id': document_id,
                'status': 'success',
                'message': 'Document processed successfully',
                'metadata': result.get('metadata', {}) # Return top-level metadata
            }

    except Exception as e:
        # Catch unexpected errors during the call
        logger.exception(f"Unexpected error calling DocumentProcessor for {file_path}: {e}")
        # Store minimal failure info
        documents[document_id] = {
            'id': document_id,
            'file_path': file_path,
            'file_name': file_name,
            'uploaded_at': datetime.now().isoformat(),
            'processing_result': {'error': str(e)},
            'metadata': {'processing_status': 'failed'}
        }
        return {
            'document_id': document_id,
            'status': 'error',
            'message': f"Unexpected error during processing: {str(e)}",
            'metadata': {'processing_status': 'failed'}
        }

def get_all_documents():
    """
    Get all documents
    
    Returns:
        list: All documents
    """
    return [
        {
            'id': doc_id,
            'file_name': info['file_name'],
            'uploaded_at': info['uploaded_at'],
            'metadata': info['metadata']
        }
        for doc_id, info in documents.items()
    ]

def get_document_by_id(document_id):
    """
    Get a document by ID
    
    Args:
        document_id: Document ID
        
    Returns:
        dict: Document info or None if not found
    """
    if document_id in documents:
        return documents[document_id]
    return None

def delete_document(document_id):
    """
    Delete a document
    
    Args:
        document_id: Document ID
        
    Returns:
        bool: True if deleted, False if not found
    """
    if document_id in documents:
        # Delete the file if it exists
        try:
            file_path = documents[document_id]['file_path']
            if os.path.exists(file_path):
                os.remove(file_path)
        except Exception as e:
            logger.error(f"Error deleting document file: {e}")
        
        # Remove from memory
        del documents[document_id]
        return True
    
    return False

File: features/pdf_scanning/service.js
/**
 * PDF Scanning Service
 * 
 * Service layer for PDF document processing.
 * Implements business logic for the PDF scanning feature.
 */

const fs = require('fs');
const path = require('path');
const { v4: uuidv4 } = require('uuid');
const { promisify } = require('util');
const readFileAsync = promisify(fs.readFile);
const writeFileAsync = promisify(fs.writeFile);

// For a real implementation, you would use a proper database
// This is just a simple in-memory store for demonstration
const documentStore = {};

// Import the coordinator from agent framework
let agentCoordinator;
try {
  const agentFramework = require('../../agent_framework');
  agentCoordinator = agentFramework.get_coordinator();
} catch (error) {
  console.warn('Agent framework not available:', error.message);
  // Create a mock coordinator if the real one is not available
  agentCoordinator = {
    process_document: (path) => ({
      document_id: path.split('/').pop().replace('.', '_'),
      processed_at: new Date().toISOString(),
      chunks_count: 0,
      metadata: { title: 'Unknown', language: 'he', confidence: 0 }
    })
  };
}

/**
 * Process a PDF document
 * 
 * @param {string} filePath - Path to the uploaded PDF file
 * @param {Object} options - Processing options
 * @returns {Object} Processing result details
 */
async function processPdf(filePath, options = {}) {
  console.log(`Processing PDF: ${filePath} with options:`, options);
  
  const filename = path.basename(filePath);
  const documentId = filename.replace('.', '_');
  
  // In a real implementation, this would actually process the PDF
  // For now, we'll simulate processing with a delay
  await new Promise(resolve => setTimeout(resolve, 500));
  
  // Use the agent coordinator to process the document
  const processingResult = agentCoordinator.process_document(filePath, options);
  
  // Create document metadata
  const documentData = {
    metadata: {
      id: documentId,
      filename: filename,
      upload_date: new Date().toISOString(),
      status: 'completed',
      page_count: 10, // Mock data
      language: options.language || 'he',
      type: 'PDF',
      size_bytes: 1024000 // Mock data
    },
    processed_data: {
      isin_data: [
        { isin: "US0378331005", company: "Apple Inc.", market: "NASDAQ", type: "מניה" },
        { isin: "US88160R1014", company: "Tesla Inc.", market: "NASDAQ", type: "מניה" },
        { isin: "DE000BAY0017", company: "Bayer AG", market: "XETRA", type: "מניה" }
      ],
      financial_data: {
        is_financial: true,
        document_type: "דו\"ח שנתי",
        metrics: {
          assets: [
            { name: "סך נכסים", value: 1200000, unit: "₪" },
            { name: "נכסים נזילים", value: 550000, unit: "₪" }
          ],
          returns: [
            { name: "תשואה שנתית", value: 8.7, unit: "%" },
            { name: "תשואה ממוצעת 5 שנים", value: 7.2, unit: "%" }
          ],
          ratios: [
            { name: "יחס שארפ", value: 1.3 },
            { name: "יחס P/E ממוצע", value: 22.4 }
          ]
        }
      },
      tables: {
        "1": [
          {
            id: `${documentId}_table_1`,
            name: "חלוקת נכסים",
            page: 1,
            header: ["סוג נכס", "אחוז מהתיק", "שווי (₪)"],
            rows: [
              ["מניות", "45%", "450,000"],
              ["אג\"ח ממשלתי", "30%", "300,000"],
              ["אג\"ח קונצרני", "15%", "150,000"],
              ["מזומן", "10%", "100,000"]
            ]
          }
        ],
        "2": [
          {
            id: `${documentId}_table_2`,
            name: "התפלגות מטבעית",
            page: 2,
            header: ["מטבע", "אחוז מהתיק", "שווי (₪)"],
            rows: [
              ["שקל (₪)", "60%", "600,000"],
              ["דולר ($)", "25%", "250,000"],
              ["אירו (€)", "10%", "100,000"],
              ["אחר", "5%", "50,000"]
            ]
          }
        ]
      }
    }
  };
  
  // Store document data
  documentStore[documentId] = documentData;
  
  // In a real implementation, you would persist this to a database
  
  return {
    documentId,
    processingResult,
    status: 'success'
  };
}

/**
 * Get document data by ID
 * 
 * @param {string} documentId - Document ID
 * @returns {Object|null} Document data or null if not found
 */
async function getDocumentData(documentId) {
  // In a real implementation, this would query a database
  return documentStore[documentId] || null;
}

module.exports = {
  processPdf,
  getDocumentData
};
